ships/0022-build-strategy-volumes.md

```diff
@@ -0,0 +1,179 @@
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: build-strategy-volumes
+authors:
+  - "@adambkaplan"
+reviewers:
+  - "@otaviof"
+  - "@HeavyWombat"
+  - "@imjasonh"
+approvers:
+  - "@sbose78"
+  - "@SaschaSchwarze0"
+creation-date: 2021-08-18
+last-updated: 2021-08-18
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Build Strategy Volumes
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+TBD
+
+## Summary
+
+This proposal will allow build strategy authors to declare volumes which can be shared across build steps and build runs.
+The build strategy will declare the mount point for the volume which will be fixed across all build steps.
+Volumes can either have a fixed volume source, or an overridable volume source that can be set in a `Build` or `BuildRun`.
+
+## Motivation
+
+Shipwright builds have two limitations that the build APIs do not address:
+
+1. Builds cannot use persistent storage
+2. Build strategy authors cannot share information across build steps, outside of the cloned source code.
+
+Persistent storage is a core primitive that is needed to add caching mechanisms to builds.
+Caches for container image layers and dependency artifacts are commonly used to improve the performance of application builds.
+
+
+### Goals
+
+- Build strategies can declare volumes whose data can be reused across steps.
+- Developers can use persistent volumes in a `BuildRun`.
+
+### Non-Goals
+
+- Securing volume usage/access with admission control.
+- Allow developers to inject arbitrary volumes into `BuildRun`s.
+
+## Proposal
+
+### User Stories
+
+#### Share information in build steps
+
+As a build strategy author
+I want to declare volumes in my build steps
+So that I can share information across build steps
+
+#### Use persistent volumes in build runs
+
+As a developer
+I want to declare persistent volume claims in my Builds
+So that I can cache resources like image layers and artifacts for subsequent BuildRuns
+
+### Implementation Notes
+
+The `BuildStrategy` API will be enhanced so that volumes can be declared alongside the build strategy steps.
+The volumes will provide fixed mount points that will be used in all steps, along with a default source for the underlying filesystem.
+Volumes can be marked `optional` indicating that the volume is not needed for a `BuildRun` to execute.
+Volumes can also indicate if the source can be overrode by a `Build` or `BuildRun`.
+
+Both the `Build` and `BuildRun` APIs will be enhanced to allow volumes to be specified.
+The volume name referenced in the `Build` or `BuildRun` must be declared in the build strategy.
+
+#### Sample API
+
+This is a sample of the API specified above.
```

Comment:
As mentioned in the community meeting. We have [implicit emptyDir support today when a build strategy step contains volumeMounts](https://github.com/shipwright-io/build/blob/37646417e6f9d594072ae7af3d8f083425898b96/pkg/reconciler/buildrun/resources/taskrun.go#L194-L207) which we heavily use in our sample build strategies (for example [s2i](https://github.com/shipwright-io/build/blob/37646417e6f9d594072ae7af3d8f083425898b96/samples/buildstrategy/source-to-image/buildstrategy_source-to-image_cr.yaml#L17-L19) or [kaniko](https://github.com/shipwright-io/build/blob/37646417e6f9d594072ae7af3d8f083425898b96/samples/buildstrategy/kaniko/buildstrategy_kaniko_cr.yaml#L48-L50) or especially [kaniko-trivy](https://github.com/shipwright-io/build/blob/37646417e6f9d594072ae7af3d8f083425898b96/samples/buildstrategy/kaniko/buildstrategy_kaniko-trivy_cr.yaml#L44-L46)).

We have two options:

1. We retain that support. (+) we do not break our own samples nor other build strategies. (-) A build strategy author who defines a volumeMount with a misspelled volume name gets an emptyDir while (s)he expects the volume that was defined.
2. We deprecate that support, and later remove it. (+) clear API. (-) We can easily update our own samples, but break those who relied on the emptyDir magic from today.

I tend to option (2).

Reply:
Added deprecation note.